landing page: add statistics component

jennur opened the pull request:
Closes https://github.com/inveniosoftware/invenio-stats/issues/120

## Screenshots

![Screenshot 2023-02-13 at 14 26 27](https://user-images.githubusercontent.com/21052053/218470490-ffdcf489-1dc3-46b8-b777-08cc74d6ca72.png)

### Expanded
![Screenshot 2023-02-13 at 14 28 57](https://user-images.githubusercontent.com/21052053/218470842-09bd0071-8b94-4feb-acd3-8a0a236832d0.png)

### Hover info-icon
![Screenshot 2023-02-13 at 14 30 48](https://user-images.githubusercontent.com/21052053/218471215-f0738df7-6539-4211-b330-ebffda7df564.png)

### Hover compact number
![Screenshot 2023-02-13 at 14 29 02](https://user-images.githubusercontent.com/21052053/218470855-fe72929f-ccec-47d4-b588-40517376e1e4.png)

slint commented on invenio_app_rdm/records_ui/views/filters.py:
```diff
@@ -121,3 +122,24 @@ def get_scheme_label(scheme):
+    number = int(value)
+    decimals = 0
+
+    if number > 1000000:
```
```suggestion
    if number > 1_000_000:
```

minor: to avoid the "magic number" we could make this an optional parameter to the Jinja filter, so that someone overriding the template could change this.

Ditto for other places where this is hardcoded.

(also better Python number formatting :) )

kpsherva commented:
Looks nice! How does it look like on mobile?

slint reviewed (commented):
Look great! Minor comment on the Jinja filter definition for some possible reuse/customization if needed.

jennur commented:
## Updates

- Made jinja filters more dynamic
- Added better Python number formatting
- Added data-labels for mobile table

### Mobile layout
<img width="448" alt="Screenshot 2023-02-15 at 15 07 50" src="https://user-images.githubusercontent.com/21052053/219053234-c571c66e-5629-45fc-951a-84e1ec545911.png">

Pineirin commented:
For me the mobile version looks a bit odd. I think it has to do with the fact that is not centered.

Could you share a screenshot with it centered (sorry if it was shown in the demo, I don't remember it)?

Pineirin commented on invenio_app_rdm/records_ui/views/records.py:
```diff
@@ -100,6 +100,22 @@ def record_detail(pid_value, record, files, is_preview=False):
         custom_fields_ui=load_custom_fields()["ui"],
         is_preview=is_preview,
         is_draft=is_draft,
+        stats={
+            "this_version": {
+                "unique_views": "1274245",
+                "views": "14906489",
+                "unique_downloads": "72009789675",
+                "downloads": "832228",
+                "data_volume": "5000000000",
+            },
+            "all_versions": {
+                "unique_views": "1127502",
+                "views": "149081197",
+                "unique_downloads": "72009237812",
+                "downloads": "832228",
+                "data_volume": "5000000000000",
+            },
+        },
```
Question. Why is this hardcoded?

jennur replied:
The backend is not ready yet. It's just a draft :)

max-moser replied:
The actual statistics can be queried like this:
```python

def get_query(query_name):
    # TODO this could be cached in the invenio_stats extension
    from invenio_stats.proxies import current_stats

    query_config = current_stats.queries[query_name]
    query = query_config.cls(name=query_name, **query_config.params)

    return query


def get_stats(record_ui):
    # TODO this could be cached
    breakpoint()
    recid = record_ui["id"]
    parent_recid = record_ui["parent"]["id"]

    views = get_query("record-view").run(recid=recid)
    views_all = get_query("record-view-all-versions").run(parent_recid=parent_recid)
    downloads = get_query("record-download").run(recid=recid)
    downloads_all = get_query("record-download-all-versions").run(
        parent_recid=parent_recid
    )

    stats = {
        "this_version": {
            "views": views["count"],
            "unique_views": views["unique_count"],
            "downloads": downloads["count"],
            "unique_downloads": downloads["unique_count"],
            "data_volume": downloads["volume"],
        },
        "all_versions": {
            "views": views_all["count"],
            "unique_views": views_all["unique_count"],
            "downloads": downloads_all["count"],
            "unique_downloads": downloads_all["unique_count"],
            "data_volume": downloads_all["volume"],
        },
    }

    return stats
```

Then the hard-coded `stats={...}` could be replaced with `stats=get_stats(record_ui)`

zzacharo commented:
> For me the mobile version looks a bit odd. I think it has to do with the fact that is not centered.
> 
> Could you share a screenshot with it centered (sorry if it was shown in the demo, I don't remember it)?

Hello! I would agree with @Pineirin that something seems a bit off but I would try to have still 2 columns if possible instead of centering the whole content :) Actually, I saw that you changed the format of showing the information. Was it not possible to keep the desktop layout? The white space in the mobile layout is it indeed there on mobile screens?

jennur commented:
> > For me the mobile version looks a bit odd. I think it has to do with the fact that is not centered.
> > Could you share a screenshot with it centered (sorry if it was shown in the demo, I don't remember it)?
> 
> Hello! I would agree with @Pineirin that something seems a bit off but I would try to have still 2 columns if possible instead of centering the whole content :) Actually, I saw that you changed the format of showing the information. Was it not possible to keep the desktop layout? The white space in the mobile layout is it indeed there on mobile screens?

I see what you mean. I will try to change the alignment, according to this the numbers should actually be right aligned (https://uxpickle.com/ux-best-practice-for-data-table-cell-content-alignment/), which I think will look better as well. The mobile version is following our default table layout. I don't think we should keep the desktop layout for mobile, it will be too smashed on smaller screens.

jennur commented:
## Updated mockups
- Right align numbers (https://uxpickle.com/ux-best-practice-for-data-table-cell-content-alignment/)

### Desktop
<img width="488" alt="Screenshot 2023-02-23 at 14 52 16" src="https://user-images.githubusercontent.com/21052053/220928225-06713420-9a63-4886-acb2-2b44a6c54487.png">

### Tablet/mobile
<img width="399" alt="Screenshot 2023-02-23 at 14 53 05" src="https://user-images.githubusercontent.com/21052053/220928400-1ec05b17-7372-4e55-bf91-bb8429bf4df1.png">

kpsherva commented on invenio_app_rdm/theme/assets/semantic-ui/less/invenio_app_rdm/theme/collections/table.overrides:
```diff
@@ -58,6 +58,11 @@
               background-color: lighten(@borderColor, 75%);
             }
 
+            &.right.aligned {
```
to be tested: does this break any other tables?

kpsherva commented on invenio_app_rdm/theme/views.py:
```diff
@@ -32,6 +32,7 @@ def create_blueprint(app):
     blueprint.add_url_rule(routes["index"], view_func=index)
     blueprint.add_url_rule(routes["robots"], view_func=robots)
     blueprint.add_url_rule(routes["help_search"], view_func=help_search)
+    blueprint.add_url_rule(routes["help_statistics"], view_func=help_statistics)
```
shelve, for general discussion: I am not sure if having `APP_RDM_ROUTES` config variable to configure urls is the best

kpsherva commented on invenio_app_rdm/theme/templates/semantic-ui/invenio_app_rdm/help/statistics.en.html:
```diff
@@ -0,0 +1,198 @@
+{%- extends config.BASE_TEMPLATE %}
+
+{%- block page_body %}
+  <!-- TODO: Update this page with RDM specific content (copied from Zenodo) -->
```
TODO: replace all Zenodo mentions on this page to make it mergeable

ppanero commented on invenio_app_rdm/records_ui/views/filters.py:
```diff
@@ -121,3 +122,31 @@ def get_scheme_label(scheme):
+
+def localize_number(value):
+    """Format number according to locale value."""
+    locale_value = current_app.config.get("BABEL_DEFAULT_LOCALE")
```
note to self: this is always defined (has a default, https://github.com/inveniosoftware/invenio-app-rdm/blob/master/invenio_app_rdm/config.py#L122)

ppanero commented on invenio_app_rdm/theme/templates/semantic-ui/invenio_app_rdm/help/statistics.en.html:
```diff
@@ -0,0 +1,197 @@
+{#
+  Copyright (C) 2021 CERN.
```
```suggestion
  Copyright (C) 2021-2023 CERN.
```
